laurent22/joplin
# Mobile: Add sync status icon

# Summary
 * [Original PR](https://github.com/laurent22/joplin/pull/6785)
 * [Feature discussion](https://discourse.joplinapp.org/t/mobile-display-sync-status-while-in-edit-view/27034/4)
 * Adds an icon button that shows whether changes made in the mobile editor haven't been synced to the server.

# Testing plan
0. Select a sync target and enable the sync icon in settings > Note > Show sync status.
1. In "configuration", disable syncing
2. Edit a note.
    - The "sync status" icon should be invisible
3. In "configuration", set a cloud service as the sync target
4. Open a note.
    - The sync icon should be a "push" icon before changes are made/have been pushed to the server 
5. Push the "sync" button
    - After syncing, the icon should contain a checkmark.
6. Edit the note
    - After making a change, but before syncing, the button should contain a "push icon" again 
7. Press the "sync" button
8. In "configuration", enable syncing with the filesystem, but with an empty/invalid output directory
9. Open a note and press the "sync" button.
    - After pressing the button once, the cloud should contain an exclamation point.
    - Attempting to sync again should display an error message.

# Notes

<strike>

 * The feature is opt-in and the setting is only visible when a sync target has been selected.]
     * This selection was made as a result of [this poll](https://discourse.joplinapp.org/t/mobile-display-sync-status-while-in-edit-view/27034/11)

</strike>

 * This feature is enabled by default and cannot be disabled.


# Screenshots
![Screenshot of the "show sync status" setting under note.](https://user-images.githubusercontent.com/46334387/188492697-5b53e044-d33a-443a-938d-a5ddcc5901d9.jpg)

 - When up-to-date
     - ![Outlined cloud with a checkmark](https://user-images.githubusercontent.com/46334387/187069514-e276a3df-fa98-49b4-958e-a3e3d486c753.jpg)

 - While syncing
     - ![Outlined cloud with a 'syncing' symbol](https://user-images.githubusercontent.com/46334387/187069584-91b46c39-f531-4e60-ac44-caea330eb6cf.jpg)

 - When changes can be pushed
     - ![Outlined cloud with an upward-pointing arrow](https://user-images.githubusercontent.com/46334387/187069623-1e1f178b-5e91-4ecb-926e-55106aaadde1.jpg)

 - After sync errors
     - ![Outlined cloud containing an exclamation point](https://user-images.githubusercontent.com/46334387/187069659-1f447a7b-6fbc-4a02-8365-60b137dd3e29.jpg)
